camel/agents/chat_agent.py

```diff
@@ -78,6 +79,31 @@ def to_openai_message(self):
         return self.message.to_openai_message(self.role_at_backend)
 
 
+@dataclass(frozen=True)
+class FuncRecord:
+    r"""Historical records of functions called in the conversation.
+
+    Attributes:
+        func_name (str): Name of the function being called
+        args (Dict[str, Any]): Dictionary of arguments passed to the function
+        result (Any): The execution result of calling this function
+    """
+    func_name: str
+    args: Dict[str, Any]
+    result: Any
+
+    def __str__(self) -> str:
+        r"""Overriden version of string function
+
+        Returns:
+            str: Modified string to be represent the function calling
+        """
+
+        return (f"Function Execution: {self.func_name}\n"
+                f"\tArgs: {self.args}\n"
+                f"\tResult: {self.result}")
+
+
 class ChatAgent(BaseAgent):
     r"""Class for managing conversations of CAMEL Chat Agents.
 
```

Comment:
> In general, Any should not be there. Here we could make an empty ABC base class for `ChatGPTConfig` and other model configs. But a better solution would be injecting a pre constructed model backend. This will replace `model` and `model_config` and improve code cohesion. @lightaime

Reply:
@Obs01ete I have added an empty ABC class called `BaseConfig` in `configs.py` and modified the argument type to be `Optional[BaseConfig]`. I am wondering whether this is OK.